### PR TITLE
Make child fragments retrieve all shared data from the activity using RequestFutures.

### PR DESCRIPTION
**Pull request by kenzshelley**
<h1> Yet another pr that doesn't actually make visible changes. </h1>

<h2> What and Why? </h2>

This PR updates the way that we handle accessing pieces of data that are shared by multiple fragments. 
- previously, if a fragment needed something (say a spot), it would either send the request to get the spot itself, or trust that the activity already knew what the spot was (depending on which state the fragment corresponded to). 
- This was a shitty system -- it was hard to follow, and meant that multiple fragments were making the same requests.
- Now any shared resource is managed by the activity. A fragment can just say: 

```
mSpot = mCallback.getSpot(spotId).get();
```

to retrieve a resource from the activity. get[Resource] returns a RequestFuture, which essentially acts like a promise. The call to '.get()' will wait until the Future resolves, and then return it's value. This is a blocking call, thus it must be done in a separate thread. 

Note that when the activity returns a resource, it first checks to see if it already has the resource; if it doesn't, it makes the request to the backend to retrieve it. 

<h2> Other stuff </h2>

I also made the interfaces we use more clear, and added a few 'global' interfaces that multiple fragments implement. 

@nickcharles @matthewgrossman

### CODE REVIEW

**Comment by matthewgrossman** [comment]
I should respond to this at some point.

tomorrow I Promise()

**Comment by nickcharles** [comment]
This is sad.

**Comment by matthewgrossman** [comment]
sorry we r useless

**Comment by matthewgrossman** [comment]
oops

**Comment by matthewgrossman** [comment]
I'm sure it was a wonderful and amazin fix by the amazing @kenzshelley

**Comment by kenzshelley** [comment]
lol
